File: src/wheeltec_gps/wheeltec_gps_path/src/gps_path.cpp

#include "rclcpp/rclcpp.hpp"
#include <sensor_msgs/msg/nav_sat_fix.hpp>
#include <geometry_msgs/msg/pose_stamped.hpp>
#include <nav_msgs/msg/path.hpp>

#include <math.h>
#define EARTH_RADIUS 6378.137
using std::placeholders::_1;
bool pose_init;

class GpsPath : public rclcpp::Node
{
  public:
    GpsPath()
    : Node("GpsPath")
    {
      state_pub_ = create_publisher<nav_msgs::msg::Path>("gps_path", 10);
      subscription_ = this->create_subscription<sensor_msgs::msg::NavSatFix>(
      "/fix", 1, std::bind(&GpsPath::gps_callback, this, _1));
    }
    struct lla_pose
    {
      double latitude;
      double longitude;
      double altitude;
    };
    double rad(double d) 
    {
    	return d * 3.1415926 / 180.0;
    }
  
  private:
    void gps_callback(const sensor_msgs::msg::NavSatFix::SharedPtr gps_msg) 
    {
      if(!pose_init)
       {
        init_pose.latitude = gps_msg->latitude;
        init_pose.longitude = gps_msg->longitude;
        init_pose.altitude = gps_msg->altitude;
        pose_init = true;
        }
      else
        {
        //计算相对位置
            double radLat1 ,radLat2, radLong1,radLong2,delta_lat,delta_long,x,y;
            
            radLat1 = rad(init_pose.latitude);
            radLong1 = rad(init_pose.longitude);
            
            radLat2 = rad(gps_msg->latitude);
            radLong2 = rad(gps_msg->longitude);
                //计算x
            delta_lat = radLat2 - radLat1;
            delta_long = 0;
            
            
            if(delta_lat>0)
              x = -2*asin( sqrt( pow( sin( delta_lat/2 ),2) + cos( radLat1 )*cos( radLat2)*pow( sin( delta_long/2 ),2 ) ));
            else
              x = -2*asin( sqrt( pow( sin( delta_lat/2 ),2) + cos( radLat1 )*cos( radLat2)*pow( sin( delta_long/2 ),2 ) ));
                x = x*EARTH_RADIUS*1000;
        
                //计算y
            delta_lat = 0;
                delta_long = radLong2  - radLong1;
            if(delta_long>0)
              y = 2*asin( sqrt( pow( sin( delta_lat/2 ),2) + cos( radLat2 )*cos( radLat2)*pow( sin( delta_long/2 ),2 ) ) );
            else
              y = -2*asin( sqrt( pow( sin( delta_lat/2 ),2) + cos( radLat2 )*cos( radLat2)*pow( sin( delta_long/2 ),2 ) ) );
               y = y*EARTH_RADIUS*1000;
    
            //计算z
            double z = gps_msg->altitude - init_pose.altitude;
    
            //发布轨迹
            ros_path_.header.frame_id = "path";
            ros_path_.header.stamp = rclcpp::Node::now();  
            geometry_msgs::msg::PoseStamped pose;
            pose.header = ros_path_.header;
            pose.pose.position.x = x;
            pose.pose.position.y = y;
            pose.pose.position.z = z;
            ros_path_.poses.push_back(pose);
            
            state_pub_->publish(ros_path_);
    
            RCLCPP_INFO(this->get_logger(),"( x:%0.6f ,y:%0.6f ,z:%0.6f)",x ,y ,z );
    
            //
        }        
        
        
    }
    
    nav_msgs::msg::Path ros_path_;
    lla_pose init_pose;
    rclcpp::Publisher<nav_msgs::msg::Path>::SharedPtr state_pub_;
    rclcpp::Subscription<sensor_msgs::msg::NavSatFix>::SharedPtr subscription_;
};

int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  pose_init = false;
  rclcpp::spin(std::make_shared<GpsPath>());
  rclcpp::shutdown();
  return 0;
}

File: src/wheeltec_gps/nmea_navsat_driver/src/libnmea_navsat_driver/nodes/gps_path.py
import math
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import NavSatFix
from geometry_msgs.msg import PoseStamped
from nav_msgs.msg import Path

EARTH_RADIUS = 6378.137

class GpsPath(Node):
    def __init__(self):
        super().__init__('GpsPath')
        self.pose_init = False
        self.state_pub_ = self.create_publisher(Path, 'gps_path', 10)
        self.subscription_ = self.create_subscription(
            NavSatFix,
            '/fix',
            self.gps_callback,
            1
        )

    def rad(self, d):
        return d * math.pi / 180.0

    def gps_callback(self, gps_msg):
        if not self.pose_init:
            self.init_pose = self.lla_pose()
            self.init_pose.latitude = gps_msg.latitude
            self.init_pose.longitude = gps_msg.longitude
            self.init_pose.altitude = gps_msg.altitude
            self.pose_init = True
        else:
            radLat1 = self.rad(self.init_pose.latitude)
            radLong1 = self.rad(self.init_pose.longitude)
            radLat2 = self.rad(gps_msg.latitude)
            radLong2 = self.rad(gps_msg.longitude)
            
            delta_lat = radLat2 - radLat1
            delta_long = 0
            
            if delta_lat > 0:
                x = -2 * math.asin(math.sqrt(
                    math.pow(math.sin(delta_lat / 2), 2) +
                    math.cos(radLat1) * math.cos(radLat2) *
                    math.pow(math.sin(delta_long / 2), 2)
                ))
            else:
                x = -2 * math.asin(math.sqrt(
                    math.pow(math.sin(delta_lat / 2), 2) +
                    math.cos(radLat1) * math.cos(radLat2) *
                    math.pow(math.sin(delta_long / 2), 2)
                ))
            x *= EARTH_RADIUS * 1000
            
            delta_lat = 0
            delta_long = radLong2 - radLong1
            if delta_long > 0:
                y = 2 * math.asin(math.sqrt(
                    math.pow(math.sin(delta_lat / 2), 2) +
                    math.cos(radLat2) * math.cos(radLat2) *
                    math.pow(math.sin(delta_long / 2), 2)
                ))
            else:
                y = -2 * math.asin(math.sqrt(
                    math.pow(math.sin(delta_lat / 2), 2) +
                    math.cos(radLat2) * math.cos(radLat2) *
                    math.pow(math.sin(delta_long / 2), 2)
                ))
            y *= EARTH_RADIUS * 1000
            
            z = gps_msg.altitude - self.init_pose.altitude
            
            ros_path_ = Path()
            ros_path_.header.frame_id = 'path'
            ros_path_.header.stamp = self.get_clock().now().to_msg()
            pose = PoseStamped()
            pose.header = ros_path_.header
            pose.pose.position.x = x
            pose.pose.position.y = y
            pose.pose.position.z = z
            ros_path_.poses.append(pose)
            
            self.state_pub_.publish(ros_path_)
            
            self.get_logger().info("( x:%0.6f ,y:%0.6f ,z:%0.6f)", x, y, z)
    
    class lla_pose:
        def __init__(self):
            self.latitude = 0.0
            self.longitude = 0.0
            self.altitude = 0.0

def main(args=None):
    rclpy.init(args=args)
    gps_path = GpsPath()

    rclpy.spin(gps_path)
    rclpy.shutdown()

if __name__ == '__main__':
    main()